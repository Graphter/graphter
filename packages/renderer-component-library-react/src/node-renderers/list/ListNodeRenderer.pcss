@import '@graphter/theme/src/colours.pcss';

.listNodeRenderer{

}
.defaultItemWrapper{
    background-color: rgb(248, 248, 248);
    border-radius: 0.2em;
    padding: 10px;
    margin-bottom: 10px
}
.defaultNewItemWrapper{
    border: 1px solid rgb(234, 234, 234);
    border-radius: 0.2em;
    padding: 10px;
}
.items{

}
.button{
    display:inline-block;
    padding: 1em;
    border-radius: 0.5em;
    font-weight:400;
    color:var(--dark-text);
    background-color:#efefef;
    text-align:center;
    position:relative;
    border: 0 none transparent;
}