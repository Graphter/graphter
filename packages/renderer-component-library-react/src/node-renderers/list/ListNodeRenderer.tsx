import React, { ComponentType, useState } from 'react'
import { NodeRendererProps } from "@graphter/core";
import { nodeRendererStore, createDefault } from "@graphter/renderer-react";
import { useArrayNodeData } from "@graphter/renderer-react";
import { setupNodeRenderer } from "@graphter/renderer-react";
import DefaultItemView from "./DefaultItemView";
import DefaultNewItemWrapper from "./DefaultNewItemWrapper";
import DefaultEditItemWrapper from "./DefaultEditItemWrapper";
import { pathUtils } from "@graphter/renderer-react";
import { useTreeDataInitialiser } from "@graphter/renderer-react";

const ListNodeRenderer: ComponentType<NodeRendererProps> = setupNodeRenderer((
  {
    config,
    originalTreeData,
    committed,
    globalPath,
    ErrorDisplayComponent,
    options
  }: NodeRendererProps
) => {
  if (!config) throw new Error(`<ListNodeRenderer /> component at '${globalPath.join('/')}' is missing config`)
  const originalNodeData = pathUtils.getValue(originalTreeData, globalPath.slice(2), createDefault(config, []))
  if (!Array.isArray(originalNodeData)) throw new Error(`'${config.type}' renderer only works with arrays but got '${typeof originalNodeData}'`)
  if (!config.children || !config.children.length) throw new Error(`'${config.type}' renderer must have at least one child config`)
  if (config.children.length > 1) throw new Error('Only one child list type is currently supported')

  const [ showNewItemUI, setShowNewItemUI ] = useState(false)
  const [ editingItems, setEditingItems ] = useState<Set<string>>(new Set())

  const childConfig = config.children[0]
  const childRendererRegistration = nodeRendererStore.get(childConfig.type)
  if (!childRendererRegistration) throw new Error(`Couldn't find a renderer for child renderer type ${childConfig.type} at ${globalPath.join('/')}/${childConfig.id}`)
  const ChildTypeRenderer = childRendererRegistration.Renderer

  const treeDataInitialiser = useTreeDataInitialiser()

  const {
    childIds,
    removeItem,
    commitItem,
  } = useArrayNodeData(globalPath)

  return (
    <div className='flex flex-col' data-nodetype='list' data-nodepath={globalPath.join('/')}>
      <div className='' data-testid='items'>
        {childIds && childIds.map((childId: any, i: number) => {
          const childPath = [ ...globalPath, i ]
          if (!editingItems.has(childId)) return (
            <DefaultItemView
              key={childId}
              childId={childId}
              config={childConfig}
              globalPath={childPath}
              onSelect={() => {
                if (config.options?.itemSelectionBehaviour === 'INLINE' || !config.options?.itemSelectionBehaviour) {
                  const newEditingItems = new Set(editingItems)
                  newEditingItems.add(childId)
                  setEditingItems(newEditingItems)
                  return
                }
                if (config.options?.itemSelectionBehaviour === 'CUSTOM' && typeof options?.customItemSelectionBehaviour === 'function') {
                  options.customItemSelectionBehaviour(config.options?.itemSelectionBehaviour, childConfig, childPath)
                }
              }}
              onRemove={() => {
                removeItem(i)
              }}
            />
          )
          return (
            <DefaultEditItemWrapper
              key={childId}
              onRemove={() => {
                removeItem(i)
              }}
              onDone={() => {
                const newEditingItems = new Set(editingItems)
                newEditingItems.delete(childId)
                setEditingItems(newEditingItems)
              }}
            >
              <ChildTypeRenderer
                config={childConfig}
                globalPath={childPath}
                committed={committed}
                originalTreeData={originalTreeData}
                ErrorDisplayComponent={ErrorDisplayComponent}
                options={childRendererRegistration.options}
              />
            </DefaultEditItemWrapper>
          )
        })}
      </div>
      {showNewItemUI ? (
        <DefaultNewItemWrapper
          config={config}
          onAdd={() => {
            setShowNewItemUI(false)
            commitItem(childIds.length)
          }}
          onCancel={() => {
            setShowNewItemUI(false)
          }}
        >
          <ChildTypeRenderer
            config={childConfig}
            globalPath={[ ...globalPath, childIds.length ]}
            committed={false}
            originalTreeData={originalTreeData}
            ErrorDisplayComponent={ErrorDisplayComponent}
            options={childRendererRegistration.options}
          />
        </DefaultNewItemWrapper>
      ) : (
        <button
          type='button'
          className='p-5 border border-dashed rounded hover:border-blue-200 hover:bg-gray-50 transition-colours duration-200 text-blue-300'
          onClick={() => {
            treeDataInitialiser(childConfig, [ ...globalPath, childIds.length ], false)
            setShowNewItemUI(true)
          }}
          data-testid='add-item-btn'
        >[+]</button>
      )}

    </div>
  )
})

export default ListNodeRenderer