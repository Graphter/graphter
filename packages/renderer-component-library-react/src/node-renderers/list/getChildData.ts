import { GetChildDataFn } from "@graphter/core";
import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";

export const getChildData: GetChildDataFn = async (path, getNodeValue) => {
  const childIds = await getNodeValue<Array<string>>(path)
  return Promise.all(childIds.map((childId:string, i: number) => {
    const childPath = [ ...path, i ]
    const childConfig = pathConfigStore.get(childPath)
    if(!childConfig) throw new Error(`Couldn't find config for child node at path '${childPath.join('/')}'`)
    const childRenderer = nodeRendererStore.get(childConfig.type)
    return childRenderer.getChildData ?
      childRenderer.getChildData(childPath, getNodeValue) :
      getNodeValue(childPath)
  }))
}