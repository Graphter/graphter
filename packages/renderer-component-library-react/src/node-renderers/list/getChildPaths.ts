import { GetChildPathsFn, PathSegment } from "@graphter/core";
import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";

export const getChildPaths: GetChildPathsFn = async (path, getNodeValue) => {
  const childIds = await getNodeValue<Array<string>>(path)
  const results = await Promise.all(childIds.map(async (childId, i) => {
    const childPath: Array<PathSegment> = [ ...path, i ]
    const childConfig = pathConfigStore.get(childPath)
    if(!childConfig) throw new Error(`Couldn't find config for child node at path '${childPath.join('/')}'`)
    const childRenderer = nodeRendererStore.get(childConfig.type)
    if(!childRenderer.getChildPaths) return [ childPath ]
    return [ childPath, ...await childRenderer.getChildPaths(childPath, getNodeValue) ]
  }))
  return results.flat()
}