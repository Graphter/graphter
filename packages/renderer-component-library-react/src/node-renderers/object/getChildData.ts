import { GetChildDataFn } from "@graphter/core";
import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";

export const getChildData: GetChildDataFn = async (path, getNodeValue) => {
  const config = pathConfigStore.get(path)
  if (!config) throw new Error(`Couldn't find config for node at path '${path.join('/')}'`)
  if (!config.children || !config.children.length) throw new Error(`${config.type} type '${config.id}' has no children configured. At least one is required.`)

  const childValues = await Promise.all(config.children.map(childConfig => {
    const childPath = [ ...path, childConfig.id ]
    const childRenderer = nodeRendererStore.get(childConfig.type)
    const nodeDataPromise = childRenderer.getChildData ?
      childRenderer.getChildData(childPath, getNodeValue) :
      getNodeValue(childPath)
    return nodeDataPromise
      .then(value => ({ key: childConfig.id, value }))
  }))
  const reFormedObject = childValues.reduce<{ [key: string]: any }>((a, c) => {
    a[c.key] = c.value
    return a
  }, {})
  console.info(`Reformed ${childValues.length} child values for node '${config.id}'`, reFormedObject)
  return reFormedObject
}