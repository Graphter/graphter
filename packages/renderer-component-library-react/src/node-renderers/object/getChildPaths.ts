import { GetChildPathsFn } from "@graphter/core";
import { pathConfigStore, nodeRendererStore } from "@graphter/renderer-react";

export const getChildPaths: GetChildPathsFn = async (path, getNodeValue) => {
  const config = pathConfigStore.get(path)
  if(!config) throw new Error(`Couldn't find config for node at path '${path.join('/')}'`)
  if(!config.children || !config.children.length) throw new Error(`${config.type} type '${config.id}' has no children configured. At least one is required.`)
  const promises = config.children.map(async childConfig => {
    const childPath = [ ...path, childConfig.id ]
    const childRenderer = nodeRendererStore.get(childConfig.type)
    if(!childRenderer.getChildPaths) return [ childPath ]
    return [ childPath, ...await childRenderer.getChildPaths(childPath, getNodeValue) ]
  })
  const results = await Promise.all(promises)
  const paths = results.flat()
  console.info(`Found ${paths.length} paths for tree starting from '${path.join('/')}'`, paths)
  return paths
}