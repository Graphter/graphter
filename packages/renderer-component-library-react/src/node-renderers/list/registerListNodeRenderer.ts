import ListNodeRenderer from "./ListNodeRenderer";
import { NodeRendererRegistration } from "@graphter/core";
import { getChildData } from "./getChildData";
import { getChildPaths } from "./getChildPaths";

export interface ListNodeRendererOptions {
  type: string
}

export function registerListNodeRenderer(options?: ListNodeRendererOptions): NodeRendererRegistration {
  return {
    type: options?.type || 'list',
    getChildData,
    getChildPaths,
    Renderer: ListNodeRenderer,
    options
  }
}