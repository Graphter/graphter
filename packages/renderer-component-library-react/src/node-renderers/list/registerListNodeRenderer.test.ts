import { registerListNodeRenderer } from "./registerListNodeRenderer";
import { getChildData } from "./getChildData";
import { getChildPaths } from "./getChildPaths";
import ListNodeRenderer from "./ListNodeRenderer";

describe('registerListNodeRenderer()', () => {
  it('should return the correct registration', () => {
    const result = registerListNodeRenderer()
    expect(result).toMatchSnapshot()
    expect(result.getChildData).toBe(getChildData)
    expect(result.getChildPaths).toBe(getChildPaths)
    expect(result.Renderer).toBe(ListNodeRenderer)
  })
  it('should override the type if supplied in options', () => {
    const result = registerListNodeRenderer({ type: 'custom-list' })
    expect(result.type).toBe('custom-list')
  })
})
export {}