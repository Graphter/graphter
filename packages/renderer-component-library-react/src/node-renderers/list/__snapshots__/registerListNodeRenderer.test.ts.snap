// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerListNodeRenderer() should return the correct registration 1`] = `
Object {
  "Renderer": [Function],
  "description": "Manage lists of data",
  "getChildData": [Function],
  "getChildPaths": [Function],
  "name": "List",
  "options": undefined,
  "type": "list",
}
`;
